### PR TITLE
Helper functions for gismo export

### PR DESCRIPTION
**Pull request by markriegler**
# Overview
I found the current implementation of exporting gismo-compatible xml-files regarding boundary conditions, custom functions and the assembly options too cumbersome and hard to read with all the nested curly brackets. Therefore, I added three new helper functions to gismo export: one for boundary conditions, one for a function block and one for the assembly options block. This should lead to clearer code.

## New features

All related to gismo export.

* Create boundary conditions block
* Create function block
* Create assembly options block
* Add comment above one of these blocks

## Showcase
New example `examples/export_gismo.py` recreates the xml-file for gismo's Poisson example.

## Checklists
* [ ] Documentations are up-to-date.
* [x] Added example(s)
* [ ] Added test(s)

### CODE REVIEW

**Comment by j042** [comment]
Are those all new features or can this replace any of the existing section?

**Comment by markriegler** [comment]
These are all new features.

**Comment by j042** [comment]
Thanks for the update! Can you please go through the comments from the bot and mark them resolved afterwards? Feel free to leave a small comment if needed

File: examples/export_gismo.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+    additional_blocks = gismo.AdditionalBlocks()
+
+    # Create function block for rhs
+    additional_blocks.add_function_block(
+        dim=2,
+        block_id=1,
+        function_string="2*pi^2*sin(pi*x)*sin(pi*y)",
+        comment="Right-hand side function",
+    )
```
[the remark this answers is not in the document]
Comment (markriegler) [reply examples/export_gismo.py:80]: I already left a comment

File: examples/export_gismo.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+
+def create_geometry():
+    """Recreate multipatch geometry from g+smo repository
+
+    Returns
+    ---------
+    multipatch: splinepy Multipatch
+        Arc geometry consisting of two patches
+    """
```
[the remark this answers is not in the document]
Comment (markriegler) [reply examples/export_gismo.py:19]: There already is a blank line. The suggestion is the same as the original

File: examples/export_gismo.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+    arc_outer = arc_inner.copy()
+    arc_outer.control_points = [
+        [2.0, 0.0],
+        [3.0, 0.0],
+        [2.0, 2.0],
+        [3.0, 3.0],
+        [0.0, 2.0],
+        [0.0, 3.0],
+    ]
```
[the remark this answers is not in the document]
Comment (markriegler) [reply examples/export_gismo.py:43]: In my opinion not really necessary

File: examples/export_gismo.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+def generate_additional_xml_blocks():
+    """
+    Create blocks for rhs, manufactured solution, boundary conditions and assembly
+    options.
+
+    Returns
+    -----------
+    additional_blocks: list<dict>
+        List of dictionaries used for splinepy's gismo export function
+    """
```
[the remark this answers is not in the document]
Comment (markriegler) [reply examples/export_gismo.py:70]: Same as above: suggestion is the same as the code

File: examples/export_gismo.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+        function_list=[
+            "sin(pi*x) * sin(pi*y)",
+            ("pi*cos(pi*x) * sin(pi*y)", "pi*sin(pi*x) * cos(pi*y)"),
+            "0",
+        ],
+        bc_list=[("BID2", "Dirichlet", 0), ("BID1", "Neumann", 1)],
+        unknown_id=0,
+        multipatch_id=0,
+        comment="The boundary conditions (multipatch=number of patches)",
+    )
```
[the remark this answers is not in the document]
Comment (markriegler) [reply examples/export_gismo.py:103]: This suggestion is the same as my comment

File: examples/export_gismo.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+    # Create dictionary for assembly options
+    additional_blocks.add_assembly_options_block(
+        block_id=4, comment="Assembler options"
+    )
```
[the remark this answers is not in the document]
Comment (markriegler) [reply examples/export_gismo.py:108]: Same as before: suggestion is the same as my comment

File: examples/export_gismo.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+        ["Multipatch", multipatch],
+        *[
+            [
+                f"BID{i+1}: {boundary_names[i]}",
+                multipatch.boundary_multipatch(i + 1),
+            ]
+            for i in range(len(multipatch.boundaries))
+        ],
+        control_points=False,
+    )
```
[the remark this answers is not in the document]
Comment (markriegler) [reply examples/export_gismo.py:130]: Not really that necessary, in my opinion

File: tests/data/gismo_additional_blocks.xml [hunk 8]
```diff
@@ -0,0 +1,127 @@
+<xml>
```
[the remark this answers is not in the document]
Comment (markriegler) [reply tests/data/gismo_additional_blocks.xml:1]: This already happens in the 2nd line

File: tests/io/test_gismo.py [hunk 9]
```diff
@@ -263,6 +263,66 @@ def test_gismo_export_3D(
...
+            function_list=[
+                "sin(pi*x) * sin(pi*y)",
+                ("pi*cos(pi*x) * sin(pi*y)", "pi*sin(pi*x) * cos(pi*y)"),
+                "0",
+            ],
+            bc_list=[("BID2", "Dirichlet", 0), ("BID1", "Neumann", 1)],
+            unknown_id=0,
+            multipatch_id=0,
+            comment="Boundary conditions",
+        )
```
[the remark this answers is not in the document]
Comment (markriegler) [reply tests/io/test_gismo.py:300]: This is the same as my comment

File: tests/io/test_gismo.py [hunk 10]
```diff
@@ -263,6 +263,66 @@ def test_gismo_export_3D(
...
+
+        # Create dictionary for assembly options
+        additional_blocks.add_assembly_options_block(
+            block_id=4, comment="Assembler options"
+        )
```
[the remark this answers is not in the document]
Comment (markriegler) [reply tests/io/test_gismo.py:305]: Same as my comment

File: tests/io/test_gismo.py [hunk 11]
```diff
@@ -263,6 +263,66 @@ def test_gismo_export_3D(
...
+            )
+
+            with open(tmpf) as tmp_read, open(
+                os.path.dirname(__file__)
+                + "/../data/gismo_additional_blocks.xml"
+            ) as base_file:
+                assert are_stripped_lines_same(
+                    base_file.readlines(), tmp_read.readlines(), True
+                )
+
```
[the remark this answers is not in the document]
Comment (markriegler) [reply tests/io/test_gismo.py:324]: Same as above: there are already comments

File: tests/io/test_gismo.py [hunk 12]
```diff
@@ -263,6 +263,66 @@ def test_gismo_export_3D(
...
+
```
[the remark this answers is not in the document]
Comment (markriegler) [reply tests/io/test_gismo.py:325]: No. There already are blank lines after the test...

File: examples/export_gismo.py [hunk 13]
```diff
@@ -0,0 +1,141 @@
+"""
+Export to g+smo-compatible xml-file.
+
+Recreates poisson2d_bvp.xml from gismo repo
+(see https://github.com/gismo/gismo/blob/stable/filedata/pde/poisson2d_bvp.xml)
+"""
```
[the remark this answers is not in the document]
Comment (markriegler) [reply examples/export_gismo.py:6]: This keeps popping up. There already is a blank line

File: examples/export_gismo.py [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+        control_points=False,
+    )
+
+    # Export multipatch geometry and additional options/functions to an XML file
+    gismo.export(
+        fname="poisson2d_bvp_recreation.xml",
+        multipatch=multipatch,
+        indent=True,
+        options=additional_blocks,
+    )
```
[the remark this answers is not in the document]
Comment (markriegler) [reply examples/export_gismo.py:141]: This is just my comment

File: splinepy/io/gismo.py [hunk 15]
```diff
@@ -11,6 +11,264 @@
...
+            entry is the boundary name, second the type of condition (e.g.
+            "Dirichlet"), third is on which variable/unknown it should be applied
+        unknown_id : int (0)
+            ID of the unknown/variable on which this boundaryConditions block should be
+            applied
+        multipatch_id : int (0)
+            ID of the multipatch
+        comment: str
+            Comment above boundaryFunctions-block
+        """
```
[the remark this answers is not in the document]
Comment (markriegler) [reply splinepy/io/gismo.py:52]: Did not find any `if type...` in file???

File: splinepy/io/gismo.py [hunk 16]
```diff
@@ -11,6 +11,264 @@
...
+            Overhead of sparse mem. allocation: (1+bdO)(bdA*deg + bdB) [0..1]
+        qua: int
+            Number of quadrature points: quA*deg + quB
+        qub: int
+            Number of quadrature points: quA*deg + quB
+        qurule: int
+            Quadrature rule [1:GaussLegendre,2:GaussLobatto]
+        comment: str
+            Comment above OptionList-block
+        """
```
[the remark this answers is not in the document]
Comment (markriegler) [reply splinepy/io/gismo.py:197]: These descriptions are taken from g+smo's description of the individual parameters. As these may change depending on g+smo, I will leave the descriptions as they are now

File: splinepy/io/gismo.py [hunk 17]
```diff
@@ -11,6 +11,264 @@
...
+        ----------
+        dim: int
+            Physical dimension of the problem
+        id: int
+            ID of function block
+        function_string: str or tuple<str,str>
+            Function as string. If tuple is given, function will be vector-valued.
+        comment: str
+            Comment above function block
+        """
```
[the remark this answers is not in the document]
Comment (markriegler) [reply splinepy/io/gismo.py:124]: I am confused as there is no `if type...` in the code... Therefore, cannot change anything

File: examples/export_gismo.py [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+        ],
+        control_points=False,
+    )
+
+    # Export multipatch geometry and additional options/functions to an XML file
+    gismo.export(
+        fname="poisson2d_bvp_recreation.xml",
+        multipatch=multipatch,
+        indent=True,
+        options=additional_blocks,
```
[the remark this answers is not in the document]
Comment (markriegler) [reply examples/export_gismo.py]: Imo, 'additional_blocks' is more appropriate than just 'options'

[the remark this answers is not in the document]
Comment (j042) [reply examples/export_gismo.py]: I agree, so can you edit function's argument name?

[the remark this answers is not in the document]
Comment (markriegler) [reply examples/export_gismo.py]: Maybe I don't understand the question and should have been clearer, but in the latest commit the kwargs is now additional_blocks, see [my latest commit](https://github.com/markriegler/splinepy/blob/436cc25c2e41a32c5885de1871dfd9703e8608f0/examples/export_gismo.py#L140). Or do you mean something else?

[the remark this answers is not in the document]
Comment (j042) [reply examples/export_gismo.py]: Oh, I see, that's it! I was brought to an old link somehow. Thanks

**Review by j042 — APPROVED** [review]

**Comment by markriegler** [comment]
Can i merge this PR or do you do it?

**Comment by j042** [comment]
I will

**Comment by j042** [comment]
@markriegler can you comment on @s-tillmann's commits?

**Comment by markriegler** [comment]
As part of the boundary conditions in g+smo it is possible to prescribe a value at a certain corner (e.g. prescribing zero pressure at a specific corner for the lid driven cavity). @s-tillmann implemented this option for the export.
